### DISCUSSION TITLE
[KPIP-3] Support Trino(Presto SQL) Engine Design

### DISCUSSION THREAD

**Discussion by hddong** [apache/kyuubi#1455] [Ideas]
The motivation and design are similer of [[KPIP-2] Kyuubi Support Flink SQL Engine Design #1132](https://github.com/apache/incubator-kyuubi/discussions/1132)

# Trino
 [Trino](https://github.com/trinodb/trino)(Presto SQL) is a fast distributed SQL query engine for big data analytics.

# Design
Trino client interact with trino cluster. The same design of [KPIP-2](https://github.com/apache/incubator-kyuubi/discussions/1132#:~:text=we%20need%20to,and%20load%20balancing.) to start trino client. We will start an Engine container process locally, yarn container and k8s pod will be supported in the future, as its runtime, which is responsible for communicating with Kyuubi Server and Trino Cluster. 
![kyuubi-trino](https://user-images.githubusercontent.com/17537134/143544902-13e78a99-78cc-4d69-a599-0467a392ee5c.png)

We need it to implement FrontendService and BackendService on the Kyuubi Engine side.
1. The front link is still based on the Thrift protocol
2. The Trino client in BackendService will base on the HTTP protocol
![protocal](https://user-images.githubusercontent.com/17537134/143550079-898a2caf-0617-4f5a-b6ac-300bd4a51f3b.png)

# Implementation
Similar of Flink engine,  the engine maintains metadata based on the Client-side. However, different from Flink, each trino resultset has its own metadata.
We will introduce:
- TrinoThriftBinaryFrontendService: to override some methods extends from ThriftFrontendService and TCLIService
- TrinoBackendService: to specify the detailed SessionManager to TrinoSessionManager
- TrinoSessionManager: to implement the interface methods comes from SessionManager, specify detailed OperationManager and manage session lifestyle.
- TrinoSessionImpl: fo Event Log.
- TrinoOperationManager: used to manage operations
- TrinoOperation: to override some aspect methods and interface methods of AbstractOperation
- ExecuteStatement: the communicate with trino cluster 
- TrinoSqlEngine: the main class of the Trino Engine

**Comment by yaooqinn** [comment]
LGTM

**Comment by zinking** [comment]
regarding the motivation, I guess my understanding is that trino/presto is iteself one sqlserver with moderate multi-tenant support. could you expand a little bit on that, just like what flink KPIP did. @hddong
